### DISCUSSION TITLE
Definition of JSX

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#14] [Chapter2]
For me, I met JSX when I studied Reactat. It was first time. So I thought, JSX is one of the React one. But that's not true. If you thought like I did. Follow me. Let's see what it is and find out what else JSX can do.

## What is JSX
![JSX](https://github.com/honaseung/react-deep-dive/assets/53066347/dee170bb-390e-48ef-94b2-d31d5c331805)
> Common confusing is `JSX` is a part of `React`. It's not one of the `ECMAScript` neither. JSX is represented by structure of `XML` or `HTML`. But it's not only purpose of JSX. Definition of JSX (according to book) is **'Make various tranpiler can tokenize tree structure which has various attribute and convert to 'ECMAScript'.** In React, usually JSX is used in return HTML. But **it's not only for that. it's designed to can use another purpose except XML or HTML.**

## What does JSX have
<pre>I don't think it's important. Just read it and keep going.</pre>
 - JSXElement
   - JSXOpeningElement ```ex) <Component>...```
   - JSXClosingElement ```ex) ...</Component>```
   - JSXSelfClosingElement ```ex) </Component>```
   - JSXFragment ```ex) <>...</>```
   - JSXElementName
     - JSXIdentifier ```ex) <SomeComponent>``` in this case 'SomeComponent' is following `JS` identifier rules.
     - JSXNameSpacedName ```ex) <SomeComponent1:SomeComponent2:...>```
     - JSXMemberExpression ```ex) <SomeComponent1.SomeComponent2....>```
       - It's allowed to use mixing up 'JSXNameSpacedName' and 'JSXMemberExpression'. But in React, we don't use 'JSXNameSpacedName' neither 'JSXMemberExpression'.
 - JSXAttributes
   - JSXSpreadAttributes ```ex) <SomeComponent {...Attributes}>```
   - JSXAttribute
     - JSXAttributeName ```ex) <SomeComponent attr1="thing1" attr:attr2={thing2}>``` in this case 'things' are following same rules of 'JSXIdentifier' and 'JSXNameSpacedName'.
     - JSXAttributeValue ```ex) <SomeComponent attr1="thing1" attr2={thing2}>``` looks same as above. 😂I couldn't find good sample.
     - JSXElement ```ex) <SomeComponent attr={AnotherComponent}> JSX's atrribute value can be another component.
 - JSXChildren
   - JSXChild
     - JSXText ```ex) <SomeComponent>TEST</SomeComponent>```
     - JSXElement ```ex) <SomeComponent></AnotherComponent></SomeComponent>```
     - JSXFragment ```ex) <SomeComponent><></></SomeComponent>```
     - JSXChildExpression ```ex) <SomeComponent>{(()=>return </AnotherComponent>)()}</SomeCoponent>```
 - JSXStrings ```ex) <SomeComponent>TEST</SomeComponent>``` did you find 'Waldo'? 🤣Neither find good sample.

## Conclusion
How long did you take to read 'What does JSX have'? Do not over 3 min. It's not important like I said! Read definition carefully more than that.
### **JSX is not only for `XML` or `HTML`. It is designed for things more than `XML` or `HTML` convert to `ECMAScript`.**

**Comment by honaseung** [comment]
What is JSX

JSX는 리액트의 일부라는 것이 일반적인 혼동입니다. 이것은 ECM스크립트 중 하나도 아닙니다. JSX는 XML이나 HTML의 구조로 대표됩니다. 하지만 이것만이 JSX만의 목적은 아닙니다. JSX의 정의(책에 따르면)는 '다양한 트랜파일러가 다양한 속성을 가진 트리 구조를 토큰화하여 'ECM스크립트'로 변환할 수 있도록 합니다. 리액트에서는 일반적으로 JSX가 HTML를 리턴하는 용도로 사용됩니다. 하지만 이 뿐만 아닙니다. XML이나 HTML을 제외한 다른 목적을 위해서도 사용할 수 있도록 설계되었습니다.

**Reply by chaminju927** [reply]
리액트로 업무를 매일 하다보니 'JSX가 리액트의 일부가 아니다, 다른 목적을 위해서도 사용될수 있다'는 부분은 저도 잠시 잊고 있었던 것 같습니다. 좋은 자료 제공해주셔서 감사합니다 !
